pkg/extensions/tbtc/tbtc.go

```diff
@@ -604,6 +616,28 @@ func (t *tbtc) watchKeepClosed(
 	return signalChan, unsubscribe, nil
 }
 
+func (t *tbtc) shouldMonitorDeposit(
+	depositAddress string,
+) bool {
+	keepAddress, err := t.chain.KeepAddress(depositAddress)
+	if err != nil {
+		return false
+	}
+
+	members, err := t.chain.GetMembers(common.HexToAddress(keepAddress))
+	if err != nil {
+		return false
+	}
+
+	for _, member := range members {
+		if member == t.chain.Address() {
+			return true
+		}
+	}
+
+	return false
```

Comment:
This function executes two calls for every state transition we are monitoring. It means we add 3x2 calls for every new deposit now and we'll have 4x2 calls once we implement provide redemption proof. So 8 calls in total for every new deposit in the system. It's not the end of the world, but it can increase operational costs.

I think we have two options to improve it: 
- use `Registry` for all monitorings but retrieve pubkey,
- introduce time cache in a similar fashion as in the firewall.

I like the second idea better because we do not depend on the order of events and when keep is added to the registry. No logic is leaking from other modules.

For the second option, we would have two time caches - one for deposits we are tracking and another for deposits we are not tracking. Two calls to the chain would be executed for every new deposit but only one time.

Reply:
I'm not sure a time cache will improve the situation. The major drawback is the fact we don't know when the deposit state will change. It may take months until the deposit redempton will be started. So, we can't determine the optimal caching timespan. Also it's not true we will have 4x2 once we implement proofs. Proofs will be in the same monitoring routine as increase redemption fee.

Reply:
Correct. Time cache would not work between deposit and redemption but we can reduce the number of calls during them separately. 

Looking at the next PR, #591, in `monitorAndAct` we do:
- `shouldMonitorFn`?
- `acquireMonitoringLock`

In this setup, we'll execute keep memebrs check for every deposit state change in the system. And we'll do it not only one time but for every event duplicate. 

One option could be to swap those calls, but then we'll be still executing keep members check for:
- all deposits we are not a part of, for which event duplicates occured and we already released the monitoring lock,
- all increase fee events for the deposit we are already monitoring.

We can set the lifetime of this cache to the maximum time of redemption with 3 x increased fee (because why not).

Reply:
Ok. Can we do it in a separate PR to not block here? I'm currently entering the optimization land so it's a good candidate.

Reply:
Sure, I'll just keep this thread open for now.

Reply:
Addressed in the follow-up PR: https://github.com/keep-network/keep-ecdsa/pull/600